### DISCUSSION TITLE
Segmentation Fault on Flight software with I2C Driver

### DISCUSSION THREAD

**Discussion by rawinza555** [nasa/fprime#798] [Q&A] [answered]
Hi it's me again.
I'm working on better understanding and using the available driver you guys make. I have somewhat success with GPIO, SPI and UART, thanks to the RPI Demo. For the last part, I'm trying to use i2c driver to read sensor reading from ADXL 345, which I have uploaded on my github repo [here](https://github.com/rawinza555/FPrime_I2C_ADXL345). Unfortunately after I run the flight software I've got a seg fault without any other explanation. 

I cross compile my code from ubuntu 18.04 to raspberry pi. The reference for programming and reading ADXL 345 is available [here](https://www.engineersgarage.com/adxl345-accelerometer-raspberry-pi-i2c/). I know I could have just use adxl 345 library but I can see myself working on an i2c device that might not have this kind of library available so I'm trying to stick to the generic software for better practice.

The error log from R PI terminal is here

> 
> sudo ./Prototype -a "my gds ip" -p 50000 
> Hit Ctrl-C to quit

> EVENT: (121) (2:1624976234,886022) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x021D registered to port 4 slot 0
> EVENT: (121) (2:1624976234,886117) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x021E registered to port 4 slot 1
> EVENT: (121) (2:1624976234,886166) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x021F registered to port 4 slot 2
> EVENT: (121) (2:1624976234,886215) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0220 registered to port 4 slot 3
> EVENT: (121) (2:1624976234,886262) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0221 registered to port 4 slot 4
> EVENT: (121) (2:1624976234,886308) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0222 registered to port 4 slot 5
> EVENT: (121) (2:1624976234,886354) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0223 registered to port 4 slot 6
> EVENT: (121) (2:1624976234,886401) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0079 registered to port 1 slot 7
> EVENT: (121) (2:1624976234,886444) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x007A registered to port 1 slot 8
> EVENT: (121) (2:1624976234,886526) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x007B registered to port 1 slot 9
> EVENT: (121) (2:1624976234,886635) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x007C registered to port 1 slot 10
> EVENT: (121) (2:1624976234,886739) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A5 registered to port 0 slot 11
> EVENT: (121) (2:1624976234,886843) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A7 registered to port 0 slot 12
> EVENT: (121) (2:1624976234,886945) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01A8 registered to port 0 slot 13
> EVENT: (121) (2:1624976234,887048) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x008D registered to port 3 slot 14
> EVENT: (121) (2:1624976234,887149) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01F5 registered to port 5 slot 15
> EVENT: (121) (2:1624976234,887252) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01F6 registered to port 5 slot 16
> EVENT: (121) (2:1624976234,887354) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x01F7 registered to port 5 slot 17
> EVENT: (121) (2:1624976234,887456) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0259 registered to port 7 slot 18
> EVENT: (121) (2:1624976234,887551) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x025A registered to port 7 slot 19
> EVENT: (121) (2:1624976234,887596) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x025B registered to port 7 slot 20
> EVENT: (121) (2:1624976234,887638) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x025C registered to port 7 slot 21
> EVENT: (121) (2:1624976234,887733) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x025D registered to port 7 slot 22
> EVENT: (121) (2:1624976234,887834) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x025E registered to port 7 slot 23
> EVENT: (121) (2:1624976234,887937) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0169 registered to port 2 slot 24
> EVENT: (121) (2:1624976234,888045) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x016A registered to port 2 slot 25
> EVENT: (121) (2:1624976234,888147) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x016B registered to port 2 slot 26
> EVENT: (121) (2:1624976234,888252) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x04B0 registered to port 6 slot 27
> EVENT: (121) (2:1624976234,888355) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x04B1 registered to port 6 slot 28
> EVENT: (121) (2:1624976234,888455) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0141 registered to port 8 slot 29
> EVENT: (121) (2:1624976234,888538) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0142 registered to port 8 slot 30
> EVENT: (121) (2:1624976234,888586) DIAGNOSTIC: (CMDDISP) OpCodeRegistered : Opcode 0x0960 registered to port 9 slot 31
> EVENT: (147) (2:1624976234,888683) WARNING_HI: (PRM) PrmFileReadError : Parameter file read failed in stage 0 with record 0 and error 1
> [INFO] Opened i2cEVENT: (21) (2:1624976234,889849) DIAGNOSTIC: (RG1) RateGroupStarted : Rate group started.
> EVENT: (241) (2:1624976234,890182) DIAGNOSTIC: (RG3) RateGroupStarted : Rate group started.
> EVENT: (161) (2:1624976234,890256) DIAGNOSTIC: (RG2) RateGroupStarted : Rate group started.
> [ERROR] Failed to send framed data: 1
> [ERROR] Failed to send framed data: 1
> Segmentation fault

I have verified that i2c is open for my pi and I have a success on reading accelerometer data via python. I wonder if anyone here have had similar problem and can guide me to how to solve this.

 Thank you in advance! Hopefully it's not me accidentally deleting block driver or something again.

**Comment by Joshua-Anderson** [comment, marked as answer]
I'd recommend running the F' application with gdb or collecting a core dump from the segfault so you can get the backtrace of the crash. A segmentation fault could be a lot of things, so figuring out where in the code the fault is occurring would probably help!

As a tip, this sort of error can be caused by incorrect topology files, so double checking to make sure you have everything connected properly and initialized is a good first step as well.

**Reply by rawinza555** [reply]
Do you know which kind of standard ports an i2c driver has? It doesn't use modeler = true so I couldn't really see or double check which ports are used except from the custom made ones.

**Reply by rawinza555** [reply]
Thank you for recommend me using gdb. Turned out I used sscanf incorrectly so I changed the way to read the sensor data. Now I seem to be somewhat successful at opening, writing and reading something from i2c port. I still have to format the data correctly though. My repo is updated to the current work.

This might not be a lot to give it back to you guys but I'm going to turn this into a i2c driver tutorial/documentation. i2c is not in a rpi demo so this might be something useful for the community. I should have this done by this weekend and will share it with you guys for some feedback.
